File: Frogger/frog.cpp
#include "frog.h"

Frog::Frog(float x, float y, float z) : DynamicObject(x, y, z) {

	for (int i = 0; i < 4; i++) {
		vsres[i] = new VSResSurfRevLib();
	}
	
	lives = 5;
	gamePoints = 0;

	setupObjects();
}

Frog::~Frog(){}

void Frog::setupObjects() {
	
	float body_amb[] = { 0.05f, 0.05f, 0.0f, 1.0f };
	float body_diff[] = { 0.1f, 0.55f, 0.1f, 1.0f };
	float body_spec[] = { 0.35f, 0.65f, 0.45f, 1.0f };
	float body_shininess[] = { 0.25f };

	float eyes_diff[] = { 1.0f, 1.0f, 1.0f, 1.0f };

	//Head
	vsres[0]->createSphere(0.5f, 40);
	vsres[0]->setColor(VSResourceLib::AMBIENT, body_amb);
	vsres[0]->setColor(VSResourceLib::DIFFUSE, body_diff);
	vsres[0]->setColor(VSResourceLib::SPECULAR, body_spec);
	vsres[0]->setColor(VSResourceLib::SHININESS, body_shininess);

	//Eyes
	vsres[1]->createSphere(0.2f, 40);
	vsres[1]->setColor(VSResourceLib::DIFFUSE, eyes_diff);
	vsres[1]->setColor(VSResourceLib::SPECULAR, eyes_diff);

	//Body
	vsres[2]->createCylinder(1.0f, 0.5f, 40);
	vsres[2]->setColor(VSResourceLib::AMBIENT, body_amb);
	vsres[2]->setColor(VSResourceLib::DIFFUSE, body_diff);
	vsres[2]->setColor(VSResourceLib::SPECULAR, body_spec);
	vsres[2]->setColor(VSResourceLib::SHININESS, body_shininess);

	vsres[3]->createSphere(0.5f, 40);
	vsres[3]->setColor(VSResourceLib::AMBIENT, body_amb);
	vsres[3]->setColor(VSResourceLib::DIFFUSE, body_diff);
	vsres[3]->setColor(VSResourceLib::SPECULAR, body_spec);
	vsres[3]->setColor(VSResourceLib::SHININESS, body_shininess);
}

void Frog::render(VSShaderLib shader) {
	
	vsml->loadIdentity(VSMathLib::MODEL);

	//Head
	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0] + 0.5f, position[1] + 0.5f, position[2]);
	vsres[0]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);

	//Eyes
	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0] + 0.8f, position[1] + 0.8f, position[2] + 0.2f);
	vsres[1]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);

	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0] + 0.8f, position[1] + 0.8f, position[2] - 0.2f);
	vsres[1]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);

	//Body
	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0], position[1], position[2]);
	vsml->rotate(90.0f, 1.0f, 0.0f, 0.0f);
	vsres[2]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);

	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0], position[1], position[2] + 0.5f);
	vsres[3]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);

	vsml->pushMatrix(VSMathLib::MODEL);
	vsml->translate(position[0], position[1], position[2] - 0.5f);
	vsres[3]->render(shader);
	vsml->popMatrix(VSMathLib::MODEL);
}

void Frog::move(float x, float y, float z, float speed) {
	
	if (!(position[0] + x*speed <= 32 && position[0] + x*speed >= 0))
		return;
	if (!(position[2] + z*speed <= 100 && position[2] + z*speed >= 0))
		return;

	position[0] += x * speed;
	position[1] += y * speed;
	position[2] += z * speed;
}

float** Frog::getBoundingBox(){

	float** boundingBox = 0;

	boundingBox = new float*[2];
	boundingBox[0] = new float[3];
	boundingBox[1] = new float[3];
	
	boundingBox[0][0] = position[0] + 1.2f;
	boundingBox[0][1] = position[1];
	boundingBox[0][2] = position[2] - 1.0f;

	boundingBox[1][0] = position[0] - 1.2f;
	boundingBox[1][1] = position[1];
	boundingBox[1][2] = position[2] + 1.0f;

	return boundingBox;
}

void Frog::resetLifes() {

	lives = 5;
	gamePoints = 0;
}

void Frog::collide(DynamicObject* dynamicObject) {

	if (collided){
		collided = false;
		return;
	}
	else {

		if (position[0] > 18 && position[0] < 28){
			loseLife();
			dynamicObject->resetPosition();
		}
		if (position[0] > 30){
			gamePoints += 10;
			dynamicObject->resetPosition();
		}
	}
}

int Frog::getPoints(){
	
	return gamePoints;
}

File: FroggerJS/frog.js
function Frog(x, y, z) {

	this.x = x;
	this.y = y;
	this.z = z;

	this.iX = x;
	this.iY = y;
	this.iZ = z;

	this.eye = new Sphere(0.2);
	this.frogBody = new Sphere(0.5);

	this.gamePoints = 0;
	this.lives = 0;

	this.setupObjects();
}

Frog.prototype.setupObjects = function() {

	//Body
	this.frogBody.setColor("ambient", [ 0.05, 0.05, 0.0, 1.0]);
	this.frogBody.setColor("diffuse", [0.1, 0.55, 0.1, 1.0]);
	this.frogBody.setColor("specular", [0.35, 0.65, 0.45, 1.0]);
	this.frogBody.setColor("emissive", [0.727811, 0.626959, 0.626959, 1.0]);
	this.frogBody.setColor("shininess", 0.25);

	//Eyes
	this.eye.setColor("ambient", [1.0, 1.0, 1.0, 1.0]);
	this.eye.setColor("diffuse", [1.0, 1.0, 1.0, 1.0]);
	this.eye.setColor("specular", [1.0, 1.0, 1.0, 1.0]);
	this.eye.setColor("emissive", [1.0, 1.0, 1.0, 1.0]);
	this.eye.setColor("shininess", 0);
}

Frog.prototype.render = function() {

	mat4.identity(mMatrix);
	
	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.5, this.y + 0.5, this.z]);
	this.frogBody.render(shaderProgram);

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.8, this.y + 0.8, this.z + 0.2]);
	this.eye.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.8, this.y + 0.8, this.z - 0.2]);
	this.eye.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x - 0.05, this.y, this.z]);
	mat4.scale(mMatrix, [1.0, 1.0, 2.0]);
	this.frogBody.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x, this.y, this.z + 0.5]);
	this.frogBody.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x, this.y, this.z - 0.5]);
	this.frogBody.render();

	popMatrix();	
}

Frog.prototype.renderReflection = function() {

	mat4.identity(mMatrix);
	
	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.5, -(this.y + 1.5), this.z]);
	this.frogBody.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.8, -(this.y + 1.8), this.z + 0.2]);
	this.eye.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x + 0.8, -(this.y + 1.8), this.z - 0.2]);
	this.eye.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x, -[this.y + 1.0], this.z]);
	mat4.scale(mMatrix, [1.0, 1.0, 2.0]);
	this.frogBody.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x, -(this.y + 1.0), this.z + 0.5]);
	this.frogBody.render();

	popMatrix();

	pushMatrix();

	mat4.translate(mMatrix, [this.x, -(this.y + 1.0), this.z - 0.5]);
	this.frogBody.render();

	popMatrix();	
}

Frog.prototype.move = function(x, y, z, speed){

	if (!(this.x + x*speed <= 32 && this.x + x*speed >= 0))
		return;
	if (!(this.z + z*speed <= 100 && this.z + z*speed >= 0))
		return;

	this.x += x * speed;
	this.y += y * speed;
	this.z += z * speed;
}

Frog.prototype.getBoundingBox = function(){

	var boundingBox = [[],[]];

	boundingBox[0][0] = this.x +1.2;
	boundingBox[0][1] = 0;
	boundingBox[0][2] = this.z - 1;

	boundingBox[1][0] = this.x - 1.2;
	boundingBox[1][1] = 0;
	boundingBox[1][2] = this.z + 1;

	return boundingBox;
}

Frog.prototype.resetLifes = function(){
	this.lives = 5;
	this.gamePoints = 0;
}

Frog.prototype.loseLife = function(){
	this.lives--;
}

Frog.prototype.getPoints = function(){
	return this.gamePoints;
}

Frog.prototype.getLives = function(){
	return this.lives;
}

Frog.prototype.isInitialPos = function() {

	if(this.x == this.iX && this.y == this.iY && this.z == this.iZ) {
		return true;
	} else {
		return false;
	}
}

Frog.prototype.resetPosition = function() {

	this.x = this.iX;
	this.y = this.iY;
	this.z = this.iZ;
}

Frog.prototype.collide = function() {
	if (this.x >= 31){
		this.gamePoints += 10;
		this.resetPosition();
	}
}